File: android/src/main/kotlin/io/siteplan/image_utilities/ImageUtilitiesPlugin.kt
package io.siteplan.image_utilities

import androidx.annotation.NonNull

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.media.ThumbnailUtils
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File
import java.io.FileOutputStream
import java.io.FileNotFoundException
import java.io.IOException

/** ImageUtilitiesPlugin */
class ImageUtilitiesPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "image_utilities")
    channel.setMethodCallHandler(this)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "generateThumbnail" -> generateThumbnail(call, result)
      "rotate" -> rotate(call, result)
      else -> result.notImplemented()
    }
  }

  /** 
    * Generates a thumbnail for an image, centered, with a given maximum size.
    *
    * @param call The method call
    * @param result The method result
    *
    * @return void
  */
  private fun generateThumbnail(call: MethodCall, result: Result) {
    // Get arguments
    val args = call.arguments as Map<String, Any>
    val imagePath = args["imagePath"] as String
    val targetPath = args["targetPath"] as String
    val maxSize = args["maxSize"] as Int
    val format = args["format"] as Int

    // Decode image
    val bitmap = BitmapFactory.decodeFile(imagePath)
    val width = bitmap.width
    val height = bitmap.height

    // Calculate new dimensions
    var newWidth = width
    var newHeight = height

    val shouldResize = width > maxSize || height > maxSize

    if (shouldResize) {
      val ratio = width.toFloat() / height.toFloat()
      newWidth = if (ratio > 1) maxSize else (maxSize * ratio).toInt()
      newHeight = if (ratio > 1) (maxSize / ratio).toInt() else maxSize
    }

    // Generate the thumbnail
    val thumbnail = ThumbnailUtils.extractThumbnail(bitmap, newWidth, newHeight)
    
    try {
      // Create a file to write the thumbnail to
      val resizedImage = File(targetPath)

      // Compress the thumbnail to the target format and save it
      val outputFile = FileOutputStream(resizedImage)
      when (format) {
          0 -> thumbnail.compress(Bitmap.CompressFormat.JPEG, 100, outputFile)
          1 -> thumbnail.compress(Bitmap.CompressFormat.PNG, 100, outputFile)
      }
      outputFile.close()
    } catch (error: FileNotFoundException) {
      result.error("FileNotFoundException", "File creating file", error.message)
    } catch (error: IOException) {
      result.error("IOException", "Error writing file", error.message)
    }

    result.success(null)
  }

  /** 
    * Rotates an image by a given number of degrees.
    *
    * @param call The method call
    * @param result The method result
    *
    * @return void
  */
  private fun rotate(call: MethodCall, result: Result) {
    // Get arguments
    val args = call.arguments as Map<String, Any>
    val imagePath = args["imagePath"] as String
    val targetPath = args["targetPath"] as String
    val degrees = args["degrees"] as Int
    val format = args["format"] as Int

    // Decode image
    val bitmap = BitmapFactory.decodeFile(imagePath)
    val width = bitmap.width
    val height = bitmap.height

    // Rotate the image
    val matrix = Matrix()
    matrix.postRotate(degrees.toFloat())
    val rotatedBitmap = Bitmap.createBitmap(bitmap, 0, 0, width, height, matrix, true)

    try {
      // Create a file to write the rotated image to
      val rotatedImage = File(targetPath)

      // Compress the thumbnail to the target format and save it
      val outputFile = FileOutputStream(rotatedImage)
      when (format) {
        0 -> rotatedBitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputFile)
        1 -> rotatedBitmap.compress(Bitmap.CompressFormat.PNG, 100, outputFile)
      }
      outputFile.close()
    } catch (error: FileNotFoundException) {
      result.error("FileNotFoundException", "File creating file", error.message)
    } catch (error: IOException) {
      result.error("IOException", "Error writing file", error.message)
    }

    result.success(null)
  }
}

File: ios/Classes/ImageUtilitiesPlugin.swift
import Flutter
import UIKit

public class ImageUtilitiesPlugin: NSObject, FlutterPlugin {

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "image_utilities", binaryMessenger: registrar.messenger())
    let instance = ImageUtilitiesPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    // Switch on call to check which method is being called
    switch call.method {
      case "generateThumbnail":
        generateThumbnail(call, result: result)
      case "rotate":
        rotate(call, result: result)
      default:
        result(FlutterMethodNotImplemented)
    }
  }

  /// Generates a thumbnail for an image, centered, with a given maximum size.
  private func generateThumbnail(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    // Get the arguments from the call
    let args = call.arguments as! [String: Any]
    let imagePath = args["imagePath"] as! String
    let targetPath = args["targetPath"] as! String
    let maxSize = args["maxSize"] as! CGFloat
    let format = args["format"] as! Int
    
    // Load the image from the path
    let image = UIImage(contentsOfFile: imagePath)

    // If the image does not exist, return an error
    if (image == nil) {
      result(
        FlutterError(
          code: "Image not found", 
          message: "The image at the specified path (\(imagePath)) could not be found.", 
          details: nil
        )
      )
      return
    }

    let width = image!.size.width
    let height = image!.size.height

    // Calculate the new size
    var newWidth = width
    var newHeight = height

    let shouldResize = width > maxSize || height > maxSize

    if (shouldResize) {
      let ratio = width / height
      newWidth = ratio > 1.0 ? maxSize : (maxSize * ratio).rounded(.up)
      newHeight = ratio > 1.0 ?  (maxSize / ratio).rounded(.up) : maxSize
    }

    // Create a new image context
    UIGraphicsBeginImageContext(CGSize(width: newWidth, height: newHeight))

    // Draw the image in the new context
    image!.draw(in: CGRect(x: 0, y: 0, width: newWidth, height: newHeight))

    // Get the new image from the context
    let newImage = UIGraphicsGetImageFromCurrentImageContext()

    // If new image is nil, return an error
    if (newImage == nil) {
      result(
        FlutterError(
          code: "Image context not found", 
          message: "The image context could not be found.",
          details: nil
        )
      )
      return
    }

    // End the context
    UIGraphicsEndImageContext()

    // Save the image to the target path, using the specified format (0 is JPEG and 1 is PNG)
    let data = format == 0 ? newImage!.jpegData(compressionQuality: 1.0) : newImage!.pngData()
    try! data!.write(to: URL(fileURLWithPath: targetPath))

    result(nil)
  }

  /// Rotates an image by a given number of degrees.
  private func rotate(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    // Get the arguments from the call
    let args = call.arguments as! [String: Any]
    let imagePath = args["imagePath"] as! String
    let targetPath = args["targetPath"] as! String
    let degrees = args["degrees"] as! Int
    let format = args["format"] as! Int

    // Load the image from the path
    let image = UIImage(contentsOfFile: imagePath)

    // If the image does not exist, return an error
    if (image == nil) {
      result(
        FlutterError(
          code: "Image not found", 
          message: "The image at the specified path (\(imagePath)) could not be found.", 
          details: nil
        )
      )
      return
    }

    // Calculate the new size
    let radians = CGFloat(degrees) * CGFloat.pi / 180.0
    let rotatedSize = CGRect(origin: .zero, size: image!.size).applying(
      CGAffineTransform(rotationAngle: radians)).size

    // Create a new image context
    UIGraphicsBeginImageContext(rotatedSize)

    // Draw the image in the new context
    let context = UIGraphicsGetCurrentContext()!
    context.translateBy(x: rotatedSize.width / 2, y: rotatedSize.height / 2)
    context.rotate(by: radians)
    let rect = CGRect(
      x: -image!.size.width / 2, 
      y: -image!.size.height / 2, 
      width: image!.size.width, 
      height: image!.size.height
    )
    image!.draw(in: rect)

    // Get the new image from the context
    let newImage = UIGraphicsGetImageFromCurrentImageContext()

    // If new image is nil, return an error
    if (newImage == nil) {
      result(
        FlutterError(
          code: "Image context not found", 
          message: "The image context could not be found.",
          details: nil
        )
      )
      return
    }

    // End the context
    UIGraphicsEndImageContext()

    // Save the image to the target path, using the specified format (0 is JPEG and 1 is PNG)
    let data = format == 0 ? newImage!.jpegData(compressionQuality: 1.0) : newImage!.pngData()
    try! data!.write(to: URL(fileURLWithPath: targetPath))

    result(nil)
  }
}
